File: Javascript/DFS/M-547. Number of Provinces.ts
function findCircleNum(isConnected: number[][]): number {
  const n = isConnected.length;
  const visited: boolean[] = new Array(n).fill(false);
  let provinces = 0;

  const dfs = (city: number) => {
    visited[city] = true;
    for (let i = 0; i < n; i++) {
      if (isConnected[city][i] && i !== city && !visited[i]) dfs(i);
    }
  };

  for (let i = 0; i < n; i++) {
    if (visited[i]) continue;
    provinces++;
    dfs(i);
  }

  return provinces;
}

const testData = [
  {
    isConnected: [
      [1, 1, 0],
      [1, 1, 0],
      [0, 0, 1],
    ],
    ans: 2,
  },
  {
    isConnected: [
      [1, 0, 0],
      [0, 1, 0],
      [0, 0, 1],
    ],
    ans: 3,
  },
  {
    isConnected: [
      [1, 0, 0, 1],
      [0, 1, 1, 0],
      [0, 1, 1, 1],
      [1, 0, 1, 1],
    ],
    ans: 1,
  },
];

for (const data of testData) {
  console.log(data['ans'] === findCircleNum(data['isConnected']));
}

File: Python/DFS/M-547. Number of Provinces.py
class Solution:
    @classmethod
    def findCircleNum(self, isConnected: list[list[int]]) -> int:
        n = len(isConnected)
        visited = [False for _ in range(n)]
        provinces = 0

        def dfs(city: int, isConnected: list[list[int]], visited: list[bool]) -> None:
            visited[city] = True
            for i in range(len(isConnected)):
                if isConnected[city][i] and city != i and not visited[i]:
                    dfs(i, isConnected, visited)

        for i in range(n):
            if not visited[i]:
                provinces += 1
                dfs(i, isConnected, visited)

        return provinces


testData = [
    {
        "isConnected": [
            [1, 1, 0],
            [1, 1, 0],
            [0, 0, 1],
        ],
        "ans": 2,
    },
    {
        "isConnected": [
            [1, 0, 0],
            [0, 1, 0],
            [0, 0, 1],
        ],
        "ans": 3,
    },
    {
        "isConnected": [
            [1, 0, 0, 1],
            [0, 1, 1, 0],
            [0, 1, 1, 1],
            [1, 0, 1, 1],
        ],
        "ans": 1,
    },
]

for data in testData:
    print(data["ans"] == Solution.findCircleNum(data["isConnected"]))
